### PR TITLE
Implements dry-run mode

### LINKED ISSUE #1150: Add option to dry-run Infection
When testing Infection on a big project you don't want to actually MT that project. Because it'll take a lot of time, and because you are testing Infection, but not that project.

So far I had to comment out `start()` call inside `ParallelProcessRunner`, and it works. But this interferes with testing. I have to un-comment it back to run PHPUnit, and comment it again to benchmark Infection. 

Another problem here is that the report gets filled with `E`'s. Therefore, some parts of our processing pipeline become ignored and unused. Say, we don't necessarily collect errored mutations for future logging. What if we are hitting a bottleneck if we do?

What I want is to be able to run Infection with `--dry-run[=N]`, where N is an optional random seed, not unlike that of PHPUnit. Then Infection would use a stand-in dummy process that will fail/err/escape pseudo-randomly according to the seed.

### PR DESCRIPTION
**Pull request by theofidry**
Closes #1150

### CODE REVIEW

File: src/Command/InfectionCommand.php [hunk 1]
```diff
@@ -191,6 +191,12 @@ protected function configure(): void
                 InputOption::VALUE_NONE,
                 'Debug mode. Will not clean up Infection temporary folder.'
             )
+            ->addOption(
+                'dry-run',
+                null,
+                InputOption::VALUE_NONE,
+                'Dry run. Will not clean up Infection temporary folder.'
```
Comment (sanmai) [inline src/Command/InfectionCommand.php]: Why keep it?

**Review by sanmai — APPROVED** [review]

File: src/Command/InfectionCommand.php (hunk 1 above)
Comment (theofidry) [reply src/Command/InfectionCommand.php]: Wrong copy/paste

**Comment by maks-rafalko** [comment]
Wanted to try it locally

```
gh pr checkout 1183
composer install

bin/infection --dry-run 
```

Initial tests fails

```
vendor/bin/phpunit
```

Test fails

```
  1) Infection\Tests\BenchmarkTest::test_all_the_benchmarks_can_be_executed with data set "Tracing" 


Error Output:
  ================
  PHP Warning:  file_get_contents(infection/tests/benchmark/Tracing/coverage/coverage-xml/index.xml): failed to open stream: No such file or directory in infection/vendor/thecodingmachine/safe/generated/filesystem.php on line 250
  PHP Fatal error:  Uncaught Safe\Exceptions\FilesystemException: file_get_contents(infection/tests/benchmark/Tracing/coverage/coverage-xml/index.xml): failed to open stream: No such file or directory in infection/vendor/thecodingmachine/safe/generated/Exce
  ptions/FilesystemException.php:9
  Stack trace:
```

After commenting out that tests - everything works as expected

Who is supposed to use this feature? only developers of Infection? I mean, do we need to document it?

**Comment by theofidry** [comment]
hm will check with the test run, I forgot about it and just ran it with skipping the initial tests.

>Who is supposed to use this feature? only developers of Infection? I mean, do we need to document it?

Mostly us for profiling, but I guess it can still be useful to get and idea of the number of mutations you get with your config

**Comment by theofidry** [comment]
Hm test pass. I think you need to do a `make profile` to untar the appropriate files for `BenchmarkTest`

**Comment by sanmai** [comment]
> Hm test pass. I think you need to do a `make profile` to untar the appropriate files for `BenchmarkTest`

**Comment by maks-rafalko** [comment]
> Hm test pass. I think you need to do a `make profile` to untar the appropriate files for `BenchmarkTest`

that's the point, there shouldn't be any dependency. Anyway, @sanmai fixed it :)

**Review by maks-rafalko — APPROVED** [review]

File: src/Command/InfectionCommand.php (hunk 1 above)
Comment (theofidry) [reply src/Command/InfectionCommand.php:198]: Fixed in #2436.

**Pull request merged**